### DISCUSSION TITLE
Stress tensor values extr

### DISCUSSION THREAD

**Discussion by p-rapan** [thelfer/tfel#152] [Q&A]
Hello everyone,
I am new to MFront, so excuse me in advance for things I missed in the documentation. 

A rather simple question, 
I would like to know how is it possible to extract certain values of the stress tensor (σ11,σ12,σ33 etc) and use them inside my integrator. 
My law includes a property that is changing relatively to the values of the stress tensor at each integration. 

Thanks everyone.

**Comment by thelfer** [comment]
Hi @Padrp,
You may have a look here: <https://thelfer.github.io/tfel/web/tensors.html#vector-notations-for-symmetric-tensors>
However, I must admit that I don't have a clue of how components of the tensor can be useful. Wouldn't invariants of the tensor be more interesting ?

**Reply by p-rapan** [reply]
Thank you Mr. Helfer for the fast response.

The property that I want to modify is a plastic property of the Drucker Prager criterion. The change relies on the stresses obtained after applying a rotation matrix to the Cauchy stress.
My idea is to rotate the stress tensor and instead of using a rotation matrix, I am aiming to apply the transformation only to the elements needed for my property.

**Reply by thelfer** [reply]
Sorry @Padrp, I surely miss things but I can't figure out how you could rotate only components of the stress and not the whole stress tensor. Regarding objectivity, I also can't figure out that you can use something else than the invariants...

**Reply by p-rapan** [reply]
Attached you will find what I want to do. 
Since I need to use only these three values of the σ` matrix, I am trying to create three different variables σ`31, σ`32, σ`33 (instead of a whole matrix σ`) which are the sole values needed to characterize my property tbt. The values of these components can be calculated analytically but for that I need the certain values of the stress tensor.

**Reply by thelfer** [reply]
Your formulae is not homogeneous: you add two stress values and the square of a stress...

**Reply by p-rapan** [reply]
Sorry for the mistake. All the stresses are squared inside the formula.

**Reply by thelfer** [reply]
I am not mistaken, you want:

s2 = tP . s . P

one you want sqrt((s2 * v) | (s2 * v)) = sqrt(v2 | ((s*s). v2)) = sqrt((s*s) : (v2 x v2)) with v2 = P. v with v2 = (0 0 1)

You can build (v2 x v2) with `Stensor::buildFromVectorDiadicProduct` and s*s using `syme(s*s)`

**Reply by p-rapan** [reply]
This is exactly what I want. I will follow this approach and I will post here if any issues.
Thank you Mr. Helfer!

**Reply by p-rapan** [reply]
Good morning Mr. Helfer,

I have already implemented my modifications to the existing mfront of DruckerPragerEcroLin that I found in Code Aster stable version. 
However there is something worth noting. Using the initial mfront (unmodified) in a simple model of a uniaxial test, it seems to be easily converging and gives me the same results as the one implemented with FORTRAN inside Code Aster. 
When I use the same mfront for a bit more complex model (still nothing too difficult) such as a 2D excavation, the mfront routine whatever I tried (according to FAQ - ''my newly built model does not converge'') seems to not converge at all. The FORTRAN code of Drucker Prager converges pretty fast for both cases.

Consequently my modified routine also works in the uniaxial but not in the excavation case.

**Reply by thelfer** [reply]
Hi @Padrp,

would you attach the code_aster file ?

Could you try to generate mtest files to see what is going wrong. You have to compile the behaviour with the `--@AsterGenerateMTestFileOnFailure=true` option. Because `code_aster` runs in a temporary directory, you have toplace the call to `STATE_NON_LINE` inside a try/except block and copy the mtest files in a directory somewhere that you will have access after the execution of `code_aster`. Something like (not sure of the syntax, it is just to have the idea):

```python
try:
   STATE_NON_LINE(....)
except:
   os.system("mkdir  /home/your_login/tmp/mtests")
   os.system("cp *.mtest /home/your_login/tmp/mtests")  
```

**Reply by p-rapan** [reply]
Thanks for once more responding to me.

Initially I tried `@AsterErrorReport`  and `debug` to compile my mfront. I applied it but also did not converge. 

I tried to compile with `@AsterGenerateMTestFileOnFailure=true` which gave me:
Fatal Python Error: init_fs_encoding: failed to get the Python codec of the filesystem encoding
Python runtime state: core initialized

ModuleNotFoundError: no module named 'encodings' 


[message.txt](https://github.com/thelfer/tfel/files/8261201/message.txt)
Attached is the code aster file. 
Also I attached the version of DP mfront I am using.

DruckPragEcroLin1.txt

**Reply by thelfer** [reply]
To be honest, I never really looked at the implementation of the Drucker-Prager plastic behaviour. I did make an attempt for the Drucker-Prager-Cap behaviour but never really tested it. The implementation that you attached is dubious in many ways and the reference documentation of code_aster gives a much better algorithm (https://www.code-aster.org/V2/doc/v14/fr/man_r/r7/r7.01.16.pdf)

**Reply by thelfer** [reply]
@Padrp I won't have the time to have a look at the Drucker-Prager plastic behaviour right now, but you may have a look at this thread: https://sourceforge.net/p/tfel/discussion/mechanicalbeahviours/thread/b419377c8b/?limit=25&page=1#4c38/7257/9fac/fad6/55b8/72a6/36fd/34d6/74a5/c823/9566/29e2/5034/5184/297a/0528/e166

**Reply by p-rapan** [reply]
Thank you Mr Helfer for the response.

I will spend time and come back when I have updates.

**Reply by p-rapan** [reply]
Hello Mr. Helfer,

Hope you are fine. I tried to start from scratch and work a bit in the DP model. 
Now I am able to see it working in 2D (simple triaxial and excavation case) and 3D for a simple model with pretty fast convergences. 

I also verified the 2D cases with the DP model already implemented in Fortran. 
Attached you will find my mfront. 

However in the excavation case, there is a small issue when I am trying to prestress the soil before the excavation (using an initial stage of applying stress and then projecting the results to the excavation stage). In this case it seems that my mfront does not take into account the prexisting initial stresses and it is giving me the same results as if it is not prestressed. 

If you could take a look I would appreciate. 
DPmodi.txt

**Reply by thelfer** [reply]
Hi @Padrp,

That's a very nice contribution. How about rewriting the entry of the gallery based on this (working) implementation. Would you be my co-author ?

Concerning your question, pre-stress is not taken into account because your internal state variables (i.e. the elastic strain) are not consistent. MFront recompute the stress from the elastic strain and barely never uses the stress at the beginning of the time. We plan to fix this properly, but it won't be available in `code_aster` for a while... See https://github.com/thelfer/MFrontGenericInterfaceSupport/issues/50 for the discussion.

In the meantime, you could use the trick to initialize the elastic strain from the stress in the @InitLocalVariables code block. Post-processing this variable will be broken, but it will work... Note that this assumes that the elastic material properties are constant...

```cxx
@InitLocalVariables{
  // beware, mu may already be defined by the StandardElasticity brick, I did not check
  const auto lambda = computeLambda(young, nu);
  mu = computeMu(young, nu);
  K = lambda + (2 * mu) / 3;
  //
  const auto s = deviator(sig);
  const auto pr = trace(sig) / 3;
  eel = (pr / (3*K)) * Stensor::Id() + s / 2mu;
  // after that computeElasticPrediction works as expected, so does the StandardElasticity brick...
  const auto sigel =  computeElasticPrediction();
  ...
}
```

Thomas

**Reply by p-rapan** [reply]
Hello Mr. Helfer,

Once again thank you for responding fast. This is more than helpful to my work. 
For the DP that's working, even if it's working fine and fast I want to add some work on the apex side to make it more robust and I will certainly come back with a newer version. 

Your solution to initialize strains even if I did not try yet, I don't see any reason for it not to work. It's pretty straightforward but I will test it in any case and come back. However a question arises, in the case that I would like to add an orthotropic behaviour for the elastic part, this may pose a problem. I would like to have your thoughts on how to implement something like that (orthotropic behaviour with this trick for a prestressed case).

**Reply by thelfer** [reply]
For the orthotropic case, just use `@RequireStiffnessTensor` or `@ComputeStiffnessTensor` for the stiffness tensor to be computed. Then:

```cxx
S = invert(D);
eel = S *sig.
```

**Reply by thelfer** [reply]
> Once again thank you for responding fast. This is more than helpful to my work.

You're welcome. Do not hesitate to contact me for a talk at the MFront User Day this year if you find interesting results.

**Reply by p-rapan** [reply]
Hello Mr. Helfer,

I would like to know if I can with MFront to use strains / deformations as initial conditions (just like I did with stress). 
In a stage-like model with 5-10 stages, I would like each stage to have initial conditions taken from the previous step calculations.

Is there a possible way of doing this?

**Reply by thelfer** [reply]
Hi Padrp, you have the total strain at the beginning of the time step which naturaly shall come from your previous stage, just like the stress. It is mostly a question handled by the finite element solver that you use.

**Reply by p-rapan** [reply]
Quick question Mr. Helfer,

I have developed my law and compiled it in Ubuntu, to use it under Code Aster.
Is there a possible way to compile it in Windows AND use it with CA windows version? 
I have seen the tutorial of building MFront on Windows but I am wondering if it has already been implemented with the CA windows version. 

Thanks in advance.

**Reply by thelfer** [reply]
Hi @Padrp

> Is there a possible way to compile it in Windows

Yes, without trouble

> AND use it with CA windows version?

I don't have the slightest clue about this. 

You probably need to have exactly the same compiler AND be sure that the developpers of windows version did not remove external laws support. You'd better ask ask them for more information.

**Reply by p-rapan** [reply]
Thanks, I will try it and come back with a hint on that.

**Reply by p-rapan** [reply]
Hello Mr.Helfer,
I am already on the way of developing a hardening DP criterion to implement it through Code Aster.
My routine uses same equations as the one already implemented in Code Aster and for certain values I get quite close results.

However for some values of my parameters, my 3D seems to be not converging whereas the one from Code Aster converges nicely.
I feel a bit stuck since I dont know where it comes from.
Would you mind taking a look at the DP am actually using? 

In simple models such as a triaxial in 2D, 3D works fine. But in a more complex model....

DPmodi.txt

**Reply by thelfer** [reply]
Hi @Padrp,

First, you shall normalize fp with the Young's modulus as follows:

````cxx
 fp = (q+pr*3*alpha-d)/young;
````
and change the associated derivatives accordingly.

Apart from that DP is singular at the apex, which can lead to the convergence issues. You may want to use an hyperbolic regularisation as in this example: https://thelfer.github.io/tfel/web/MohrCoulomb.html

**Reply by thelfer** [reply]
> Apart from that DP is singular at the apex, which can lead to the convergence issues. You may want to use an hyperbolic
> regularisation as in this example: https://thelfer.github.io/tfel/web/MohrCoulomb.html

To be more specific, I would try something like this (I did it rapidely, there can be mistakes everywhere !).

The yield surface is given by:

q = -3 * alpha * pr + d = -3 * alpha * (p-pa) with pa = -d / (3* alpha)

or

q^2 = 9 * alpha^2 * (p-pa)^2

This "line" in the q-p plane can be approximated by the hyperboloid of equation:

q^2 = -9 * alpha^2 * ((pr - pa)^2 - (pa -pc) ^2)

where pc is a regularisation parameter such that pa-pc shall be small

You shall probably plot it to verify that it works as expected.

Finally, you can replace fp by :

fp = (q/young)^2 + 9 * alpha^2 * (((pr - pa)/young)^2 - ((pa -pc)/young) ^2)

You may want to try it first with a numerical jacobian (set a perturbation value of 1e-8 using `@PertubationValueForNumericalJacobianComputation`).

Around the apex, there still may be some convergence issues related to the fact that the normal may oscillate. See this page for some hints how to handle that: https://thelfer.github.io/tfel/web/hosford.html

**Reply by thelfer** [reply]
The other way is to implement the algoritm of Souza Neto et al... Not difficult per se, but less extensible than the solution based on the Implicit DSL (difficult to combine with other inelastic flow, limited to isotropy, etc...).

**Reply by p-rapan** [reply]
Hello, 

I tried the solution of Souza Neto before by still I belive the a smoother apex is a better idea. 

I made some changes to the routine, considering that this is the 'smooth' apex version of a Drucker Prager. 
The two equations of the yield surfaces are:
q = p · tan β + c   (linear)
sqrt(( c - pt*tanβ)^2) + q^2) = p · tan β + c (hyperbolic) (see also attached image)


I implemented the linear one, gives me the same results as my previous routine (non convergence when 3D complex loading).
With the hyperbolic does not even run, whereas I dont see anything going wrong (Jacobian issue?).

It has become a headache but I feel am close to a fully operative Drucker Prager. 
DPmodi.txt
![hyperbolic](https://user-images.githubusercontent.com/101575904/175039972-35eee0aa-0708-4c9c-a3a8-69d34da0c108.png)

**Reply by thelfer** [reply]
1. You did not normalize fp as I suggested
2. You did not change the values of the derivatives of fp... See my remark here:

> You may want to try it first with a numerical jacobian (set a perturbation value of 1e-8 using
> @PertubationValueForNumericalJacobianComputation).

**Reply by thelfer** [reply]
@Padrp Could you attach some mtest file, just for testing. Thanks.
Thomas

**Reply by p-rapan** [reply]
I replaced the script with one using the numerical jacobian (both for parabolic and linear DP).
The linear DP works exactly as the other routines (non convergence in complex loading - this happens with certain values of parameters especially when changing the young modulus to a higher value - The DP in CA works fine with any value.).
The parabolic one does not converge even for a triaxial. 

I am sorry but I have never used mtest. I run the analyses in CA. 

Attached you will find the parabolic routine with numerical jacobian. 
DPmodi.txt

**Reply by thelfer** [reply]
@Padrp Just to be sure. By default, code_aster redefines the parameter `epsilon` using RESI_INTE_RELA (see https://www.code-aster.org/V2/doc/v11/fr/man_u/u4/u4.51.11.pdf) and the default value is 1.e-6 which is far from being enough. Would you try to change this value to 1.e-14 ?

**Reply by thelfer** [reply]
@Padrp I found several mistakes in your initial implementation. Please find attached several suggestions and enhancements. I'll try to make some tests on my side.

[DruckerPrager-v1.zip](https://github.com/thelfer/tfel/files/8960760/DruckerPrager-v1.zip)

**Reply by p-rapan** [reply]
Thanks for the time spent. I will run some tests and try the modifications and come back to you later.

**Reply by p-rapan** [reply]
I changed the routine based on your comments and it's now functioning pretty nicely.

I would like to introduce a transverse isotropic elasticity in a next step (pretty much the same way as the tutorial with orthotropic linear hardening law). 

However I arrive to the following issues:
1. I don't know how to use the 'trick' for taking into account the prestressing of the material. The prestressing I used in the isotropic case cannot be used.
2. Could I use the elastic prediction with this stiffness tensor?

I did not manage to make it converge even for a simple problem. I think I am missing something related to the ComputeStiffnessTensor command.
You will find the routine attached. 
DPmodi.txt

**Reply by thelfer** [reply]
Hi,

The following code shall work. [DP.zip](https://github.com/thelfer/tfel/files/8994648/DP.zip)

1. Just use `eel=invert(D)*sig`
2. Yes

> I did not manage to make it converge even for a simple problem. I think I am missing something related to the ComputeStiffnessTensor command.

Let me know :)

**Reply by thelfer** [reply]
@Padrp Please find attached a smoothed version that seems to give the appropriate results. Don't know if it performs better, but at least it seems to work.
[DruckerPrager-2.zip](https://github.com/thelfer/tfel/files/9016836/DruckerPrager-2.zip)

**Reply by thelfer** [reply]
@Padrp In addition, an implementation following de Souza Neto et al.

[DruckerPrager_SouzaNeto2008.zip](https://github.com/thelfer/tfel/files/9019659/DruckerPrager_SouzaNeto2008.zip)

**Reply by p-rapan** [reply]
Hello Mr. Helfer, 

Firstly thank you for taking time for these implementations. 
The smooth one which is of an interest is working for the time being just for the simple models I have (triaxial etc.), I tend to spend some more time on this one tomorrow, run more tests to try make it also converge in my excavation models. I will post my updates on that.

There are some more question risen that am dealing with right now, maybe you could give a hint:
1. Regarding tranvserse anisotropy (elasticity), the compute stiffness tensor is taken into account and does not affect the model in convergence, I am attaching a picture of the stiffness tensor I am trying to implement. 

Is the stiffness tensor giving me something like that? Its crucial in order to know how to define my shear modulus. 

2. Another rather simple question, based on the documentation  ''Second and fourth-order tensor objects in TFEL/Math'', if I want to extract certain components of the stress tensor it's going to be sth like sig(0) which will be σ11. I think I am getting it right?

3. How do I cite MFront ? 

Sorry for the long post, your help to me is beyond valuable.

**Reply by thelfer** [reply]
Hi @Padrp 

> Firstly thank you for taking time for these implementations.

You'll thank me if it ever works :)

> Is the stiffness tensor giving me something like that? Its crucial in order to know how to define my shear modulus.

Yes. You shall have a look at the documentation for details:

```bash
$ mfront --help-keyword=Implicit:@ComputeStiffnessTensor
```

Also, you may want to print it to see if you get what you want:

cout << D << endl;
exit(-1)

>  if I want to extract certain components of the stress tensor it's going to be sth like sig(0) which will be σ11. I think I am getting it right?

Yes. See this paragraph: https://thelfer.github.io/tfel/web/tensors.html#vector-notations-for-symmetric-tensors

> How do I cite MFront ?

Please cite the following paper:

@article{helfer_introducing_2015,
	title = {Introducing the open-source mfront code generator: Application to mechanical behaviours and material knowledge management within the {PLEIADES} fuel element modelling platform},
	volume = {70},
	issn = {0898-1221},
	url = {http://www.sciencedirect.com/science/article/pii/S0898122115003132},
	doi = {10.1016/j.camwa.2015.06.027},
	shorttitle = {Introducing the open-source mfront code generator},
	pages = {994--1023},
	number = {5},
	journaltitle = {Computers \& Mathematics with Applications},
	shortjournal = {Computers \& Mathematics with Applications},
	author = {Helfer, Thomas and Michel, Bruno and Proix, Jean-Michel and Salvo, Maxime and Sercombe, Jérôme and Casella, Michel},
	urldate = {2016-09-22},
	date = {2015-09},
	keywords = {Domain specific languages, Implicit integration schemes, Material knowledge management, Mechanical behaviour integration, Single crystal plasticity},
}

You may also want to cite the website: https://thelfer.github.io/tfel/web/index.html

**Reply by p-rapan** [reply]
Hello Mr. Helfer,

It has been a while working with MFront and implementing different versions of the Drucker Prager law with success. 
My question now reaches I think I thoroughly discussed topic, yet I need some guidance.

I implemented the linear hardening (and softening) DP in MFront to be used with Code Aster.
The mfront am attaching is working nicely and giving good results when I am studying hardening (positive value of hardening parameter),
but it struggles in softening (negative value),

I believe the script is okay, I am starting to think that to solve this I have to look more on the solver rather than MFront (but how??!) 

DPsoft.txt

Any comments or suggestions related to any softening implementation would help me a lot.
Have a nice day!

**Comment by thelfer** [comment]
Hi @Padrp,

> It has been a while working with MFront and implementing different versions of the Drucker Prager law with success.

That's great. You're ready for at the talk at the MFront User Day !

> but it struggles in softening (negative value),

That is pretty much expected. The standard Newton-Raphson easily fails in case of softening materials and there is pretty much nothing you can do on the behaviour side. Aside from convergence issues, strain localisation are to be expected and spurious mesh dependencies.

I know that the code_aster team is pretty fond of an approach based on second gradient regularization: 
<https://www.code-aster.org/doc/default/fr/man_r/r5/r5.04.03.pdf>. Don't know much about it, but it seems that you can use your standard consistutive equations in this framework..

Regards,
Thomas

**Reply by p-rapan** [reply]
> 
> I know that the code_aster team is pretty fond of an approach based on second gradient regularization: https://www.code-aster.org/doc/default/fr/man_r/r5/r5.04.03.pdf. Don't know much about it, but it seems that you can use your standard consistutive equations in this framework..
> 

I may be wrong but the regularization is mainly to avoid mesh dependency (?!). I am more concerned on the convergence issue.
I will post any advancements.
Thanks one more time.

**Reply by thelfer** [reply]
Indeed, but mesh dependency may also induce more numerical issues, in my experience. `code_aster` also have several arc-length algorithms. Maybe one of them is well suited for this regularization...

**Reply by p-rapan** [reply]
Dear Mr. Helfer,

I am continuously working with MFront. Thank you for the guidance throughout my work.

I may ask sth dumb, but is there any way to create a state variable for the deviatoric plastic strain (since it is the field that's concerning me) ?

**Reply by thelfer** [reply]
Hi @Padrp, there is no type associated with deviatoric symmetric tensors. You must use a "standard" symmetric tensor.
